src/test/java/spoon/test/template/SubstitutionTest.java

```diff
@@ -206,6 +210,116 @@ public CtClass apply(CtType targetType) {
+
+        // act
+        CtBlock<?> substitutedMethodBody = Substitution.substituteMethodBody(
+                targetClass, template, templateExecutableName
+        );
+
+        // assert
+        assertEquals(expectedMethodBody, substitutedMethodBody);
+    }
+
+    private static class BodyWithTemplatedInitializer extends ExtensionTemplate {
+        @Parameter
+        Integer _initializer_;
+
+        public void executable() {
+            Integer x = _initializer_;
+        }
+    }
+
+    @Test
+    public void testSubstituteStatementWithTemplatedInitializer() {
+        // contract: Given a statement with a templated initializer, substituteStatement should
+        // return a new statement with the initializer replaced with the value bound to the template
+        // parameter
+
+        // arrange
+        Factory factory = createFactoryWithTemplates();
+        CtClass<?> targetClass = factory.Class().create("TargetClass");
+
+        String templateExecutableName = "executable";
+        int templateVariableIndex = 0;
+        String templateVariableName = "s";
+        String initializerToSubstitute = "My chosen initializer";
+        CtStatement expectedStatement = factory.createLocalVariable(
+                factory.Type().stringType(), templateVariableName, factory.createLiteral(initializerToSubstitute)
+        );
+
+        StatementWithTemplatedInitializer template = new StatementWithTemplatedInitializer();
+        template._initializer_ = factory.createLiteral(initializerToSubstitute);
+
+        // act
+        CtStatement substitutedStatement = Substitution.substituteStatement(
+                targetClass, template, templateVariableIndex, templateExecutableName
+        );
+
+        // assert
+        assertEquals(expectedStatement, substitutedStatement);
+    }
+
+    private static class StatementWithTemplatedInitializer extends ExtensionTemplate {
+        TemplateParameter<String> _initializer_;
+
+        public void executable() {
+            String s = _initializer_.S();
+        }
+    }
+
+    @Test
+    public void testSubstituteFieldDefaultExpressionWithTemplatedInitializer() {
+        // contract: Give an expression with a templated initializer, substituteFieldDefaultExpression should
+        // return a new expression with the initializer replaced with the value bound to the template parameter
+
+        // arrange
+        Factory factory = createFactoryWithTemplates();
+        CtType<?> targetType = factory.Class().create("TargetClass");
+
+        String initializerToSubstitute = "My chosen initializer";
+        String templateFieldName = "s";
+        CtExpression expectedExpression =  factory.createCodeSnippetExpression(initializerToSubstitute);
+
+        ExpressionWithTemplatedInitializer template = new ExpressionWithTemplatedInitializer();
+        template._initializer_ = factory.createCodeSnippetExpression(initializerToSubstitute);
```

Comment:
Avoid using code snippet expressions when you can use something more precise (here, a literal). Currently, equality for code snippets doesn't even work (see #4024).

```suggestion
        CtExpression<String> expectedExpression =  factory.createLiteral(initializerToSubstitute);

        ExpressionWithTemplatedInitializer template = new ExpressionWithTemplatedInitializer();
        template._initializer_ = factory.createLiteral(initializerToSubstitute);
```